spring-cloud-gcp-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/integration/inbound/PubSubMessageSource.java

```diff
@@ -0,0 +1,145 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.pubsub.integration.inbound;
+
+import java.util.ArrayDeque;
+import java.util.List;
+import java.util.Map;
+
+import org.springframework.cloud.gcp.pubsub.core.subscriber.PubSubSubscriberOperations;
+import org.springframework.cloud.gcp.pubsub.integration.AckMode;
+import org.springframework.cloud.gcp.pubsub.integration.PubSubHeaderMapper;
+import org.springframework.cloud.gcp.pubsub.support.GcpPubSubHeaders;
+import org.springframework.cloud.gcp.pubsub.support.converter.ConvertedAcknowledgeablePubsubMessage;
+import org.springframework.integration.IntegrationMessageHeaderAccessor;
+import org.springframework.integration.endpoint.AbstractFetchLimitingMessageSource;
+import org.springframework.integration.endpoint.AbstractMessageSource;
+import org.springframework.integration.mapping.HeaderMapper;
+import org.springframework.messaging.Message;
+import org.springframework.util.Assert;
+
+/**
+ * A <a href="https://cloud.google.com/pubsub/docs/pull#pubsub-pull-messages-sync-java">PubSub Synchronous pull</a>
+ * implementation of {@link AbstractMessageSource}.
+ *
+ * @author Elena Felder
+ *
+ * @since 1.2
+ */
+public class PubSubMessageSource extends AbstractFetchLimitingMessageSource<Object> {
+
+	private final String subscriptionName;
+
+	private final PubSubSubscriberOperations pubSubSubscriberOperations;
+
+	private AckMode ackMode = AckMode.MANUAL;
+
+	private HeaderMapper<Map<String, String>> headerMapper = new PubSubHeaderMapper();
+
+	private Class payloadType = byte[].class;
+
+	private boolean blockOnPull;
+
+	private ArrayDeque<ConvertedAcknowledgeablePubsubMessage> cachedMessages = new ArrayDeque<>();
+
+	public PubSubMessageSource(PubSubSubscriberOperations pubSubSubscriberOperations,
+			String subscriptionName) {
+		Assert.notNull(pubSubSubscriberOperations, "Pub/Sub subscriber template can't be null.");
+		Assert.notNull(subscriptionName, "Pub/Sub subscription name can't be null.");
+		this.pubSubSubscriberOperations = pubSubSubscriberOperations;
+		this.subscriptionName = subscriptionName;
+	}
+
+	public void setAckMode(AckMode ackMode) {
+		Assert.notNull(ackMode, "The acknowledgement mode can't be null.");
+		this.ackMode = ackMode;
+	}
+
+	public void setPayloadType(Class payloadType) {
+		Assert.notNull(payloadType, "The payload type cannot be null.");
+		this.payloadType = payloadType;
+	}
+
+	public void setHeaderMapper(HeaderMapper<Map<String, String>> headerMapper) {
+		Assert.notNull(headerMapper, "The header mapper can't be null.");
+		this.headerMapper = headerMapper;
+	}
+
+	/**
+	 * Instructs synchronous pull to wait until at least one message is available.
+	 */
+	public void setBlockOnPull() {
+		this.blockOnPull = true;
```

Comment:
Would be better to make it really as a setter: `setBlockOnPull(boolean blockOnPull)`.
This way end-user won't need to have an `if...else` in case this option comes from some configuration.